### PR TITLE
Refactor primary expressions

### LINKED ISSUE #10: Typed arrays literals (eg `Int[1,2,3]`)
I get a parsing error when feeding a [typed array literal](https://docs.julialang.org/en/v1/manual/arrays/#Typed-array-literals) to emacs-tree-sitter

If I input
```julia
Float64[1, 2]
```
I get
```
((source_file
  (subscript_expression
   (identifier)
   (ERROR
    (number))
   (number)))
 . 75)
```

### LINKED ISSUE #69: Error for parameterized types with subtype
`Vector{<:Number}`
```
source_file [0, 0] - [1, 0]
  parameterized_identifier [0, 0] - [0, 16]
    identifier [0, 0] - [0, 6]
    type_argument_list [0, 6] - [0, 16]
      ERROR [0, 7] - [0, 9]
        operator [0, 7] - [0, 9]
      identifier [0, 9] - [0, 15]
```

### PR DESCRIPTION
**Pull request by savq**
- Add `_quotable` rule for expressions that can be quoted without extra parentheses and disallow arbitrary expressions in quotes and interpolations.
- Add quote and interpolation expressions to `_primary_expression`.
- Add field names to values in field, index and parametrized types expressions.
- Replace `parameterized_identifier` with `parametrized_type_expression`:
  - Remove `type_argument_list`
  - Allow any `_primary_expression` to be parameterized
  - Allow subtype clauses (unary `<:`) in type arguments
  - Allow empty type argument list (e.g. `Tuple{}`)
- Rename `subscript_expression` to `index_expression`.
- Fix parenthesized expressions allowing arbitrary blocks.
- Fix identifiers allowing `@`.
- Add more tests for primary expressions: fields, indices, curly, quotes and interpolations.
- Remove 11 files from known failures.

---

- closes #10 
- closes #69 
- closes #49. Restricting quotables fixes the parsing problem listed in the issue. However, this doesn't improve #11,  so changing the precedence of the `colon_quote` might still be necessary.

### CODE REVIEW

File: grammar.js [hunk 1]
```diff
@@ -477,92 +475,68 @@ grammar({
...
-
-    parameterized_identifier: $ => seq(
-      choice($.identifier, $.field_expression),
-      $.type_argument_list
-    ),
 
-    type_argument_list: $ => seq(
+    curly_expression: $ => seq(
+      $._primary_expression,
       '{',
```
Comment (ChrHorn) [inline grammar.js:537]:
```suggestion
      'token.immediate('{'),',
```

**Review by ChrHorn — COMMENTED** [review]
Do we really want to call it `curly_expression`? I would argue that the naming in the julia parser is mostly due to historical reasons. The `{}` syntax used to be used in other places but now is only used for [parametric types](https://docs.julialang.org/en/v1/manual/types/#man-parametric-composite-types). Just calling it `parametric_type` should be fine and would also make the parse tree more readable.

File: grammar.js (hunk 1 above)
Comment (savq) [reply grammar.js:537]: > Do we really want to call it curly_expression? 

Yeah, `curly` isn't exactly the most useful name, but it's important it ends with `_expression` like the other primary expression. Maybe `parametrized_expression`? I prefer that spelling to `parameterized_` (but both are fine).

> I would also prefer to keep type_argument_list or something similar as a scope.

I think using field names for the object would be better for this, and we can add them to all primary expression. So the output would look like:

```scheme
; xs[i, j, …]
(index_expression 
  name: (identifier)
  (idenfier)
  (idenfier)
  …)

; T{R, S, …}
(parametric_expression 
  name: (identifier)
  (idenfier)
  (idenfier)
  …)

; etc.
```

Comment (ChrHorn) [reply grammar.js:537]: 1. Think `paramerized_expression` might be a bit too general. Functions with `where` syntax would also fall under this. `parameterized_type_expression` is a bit wordy but should be fine. Could also think about dropping `_expression` altogether but this is probably better done later and in a separate discussion.

2. Using field names has the disadvantage that you lose the gradual selection/movement on the tree-sitter nodes. They are only useful for the queries. Most grammars seem to currently do it this way, so this is fine for me. We can also discuss this later in a different issue.

Comment (savq) [reply grammar.js:537]: I updated the name to `parametrized_type_expression`. I kept the "parametrized" spelling because it won't match `parameter*` when grepping.

I didn't add field names to call expressions because those already have an `argument_list`.

PS: I have to update assignments before adding the immediate brace in `parametrized_type_expression`, to avoid a conflict with `_function_signature`.
